jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java

```diff
@@ -350,14 +371,20 @@ public Runnable produce()
                         if (task != null)
                             return task;
 
-                        // If more references than 1 (ie not just us), don't refill into buffer and risk compaction.
+                        // If the application has retained the content chunks then we must not overwrite content.
                         if (networkBuffer.isRetained())
-                            reacquireNetworkBuffer();
+                        {
+                            // If there is sufficient space available, we can top up the buffer rather than allocate a new one
+                            if (minBufferSpace > 0 && BufferUtil.space(networkBuffer.getByteBuffer()) >= minBufferSpace)
+                                // do not compact the buffer
+                                compact = false;
+                            else
+                                // otherwise reacquire the buffer and fill into the new buffer.
+                                reacquireNetworkBuffer();
+                        }
                     }
 
-                    // Here we know that this.networkBuffer is not retained by
-                    // application code: either it has been released, or it's a new one.
-                    int filled = fill(getEndPoint(), networkBuffer.getByteBuffer());
+                    int filled = fill(getEndPoint(), networkBuffer.getByteBuffer(), compact);
```

Comment:
In HTTP/1 we keep the buffer around if `filled<=0`, since it's retained anyway by the application.
Should not we do the same here?